src/QsCompiler/CompilerOptimizations/Utils.fs

```diff
@@ -0,0 +1,203 @@
+﻿module Microsoft.Quantum.QsCompiler.CompilerOptimization.Utils
+
+open System
+open System.Collections.Generic
+open System.Collections.Immutable
+open System.Numerics
+open Microsoft.Quantum.QsCompiler.DataTypes
+open Microsoft.Quantum.QsCompiler.SyntaxExtensions
+open Microsoft.Quantum.QsCompiler.SyntaxTokens
+open Microsoft.Quantum.QsCompiler.SyntaxTree
+
+
+/// Option default value operator
+let (|?) = defaultArg
+
+
+/// Shorthand for a QsExpressionKind
+type Expr = QsExpressionKind<TypedExpression, Identifier, ResolvedType>
+/// Shorthand for a QsTypeKind
+type TypeKind = QsTypeKind<ResolvedType, UserDefinedType, QsTypeParameter, CallableInformation>
+
+
+/// Represents the global dictionary that maps names to callables
+type CallableDict = {
+        compiledCallables: ImmutableDictionary<QsQualifiedName, QsCallable>
+} with
+    member this.getCallable (name: QsQualifiedName): QsCallable =
+        this.compiledCallables.[name]
+
+
+/// Returns whether a given expression is a literal (and thus a constant)
+let rec isLiteral (expr: Expr) (cd: CallableDict): bool =
+    match expr with
+    | UnitValue | IntLiteral _ | BigIntLiteral _ | DoubleLiteral _ | BoolLiteral _ | ResultLiteral _ | PauliLiteral _ -> true
+    | ValueTuple a | StringLiteral (_, a) | ValueArray a -> Seq.forall (fun x -> isLiteral x.Expression cd) a
+    | RangeLiteral (a, b) -> isLiteral a.Expression cd && isLiteral b.Expression cd
+    | NewArray (_, a) -> isLiteral a.Expression cd
+    | Identifier (GlobalCallable _, _) | MissingExpr -> true
+    | CallLikeExpression ({Expression = Identifier (GlobalCallable qualName, _)}, b)
+        when (cd.getCallable qualName).Kind = TypeConstructor -> isLiteral b.Expression cd
+    | CallLikeExpression (a, b) ->
+        isLiteral a.Expression cd && isLiteral b.Expression cd &&
+            TypedExpression.IsPartialApplication (CallLikeExpression (a, b))
+    | _ -> false
+
+
+/// Helper method to improve expression readability
+let rec prettyPrint (expr: Expr): string =
```

Comment:
You could also just generate Q# for the given expression (using the ExpressionKindToQs transformation)?

Reply:
Yeah... I didn't know about the ExpressionKindToQs transformation. I expanded this function significantly in the Loop Unrolling PR, see the Printer.fs file there. An existing transformation sounds like it would have been much easier.